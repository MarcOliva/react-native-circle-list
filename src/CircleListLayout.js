import React from 'react'
import { StyleSheet, View } from 'react-native'
import PropTypes from 'prop-types'

const { cos, PI, sin } = Math

export const CircleListLayout = ({
    calcHeight,
    containerStyle,
    displayData,
    initialRotationOffset,
    keyExtractor,
    panHandlers,
    radius,
    renderItem,
    rotationOffset,
    selectedIndex,
    selectedItemScale,
    theta,
}) => (
    <View {...panHandlers} style={[styles.container, { height: calcHeight() }, containerStyle]}>
        {displayData.map((item, index) => {
            const thetaOffset = 2 * PI * index + (rotationOffset + initialRotationOffset)
            const translateX = radius * cos(index * theta + thetaOffset)
            const translateY = radius * sin(index * theta + thetaOffset) + radius
            const selected = selectedIndex === index

            return (
                translateY < radius && (
                    <View
                        key={keyExtractor(item, index)}
                        style={[
                            styles.renderItemContainer,
                            {
                                transform: [
                                    { translateX },
                                    { translateY },
                                    { scale: selected ? selectedItemScale : 1 },
                                ],
                            },
                        ]}
                    >
                        {renderItem({ item, index })}
                    </View>
                )
            )
        })}
    </View>
)

const styles = StyleSheet.create({
    container: {
        position: 'relative',
        flexDirection: 'row',
        justifyContent: 'center',
        alignSelf: 'stretch',
    },
    renderItemContainer: {
        position: 'absolute',
    },
})

CircleListLayout.propTypes = {
    calcHeight: PropTypes.func.isRequired,
    containerStyle: PropTypes.object,
    data: PropTypes.array,
    displayData: PropTypes.array,
    elementCount: PropTypes.number,
    initialRotationOffset: PropTypes.number,
    keyExtractor: PropTypes.func.isRequired,
    panHandlers: PropTypes.object.isRequired,
    radius: PropTypes.number,
    renderItem: PropTypes.func.isRequired,
    rotationOffset: PropTypes.number.isRequired,
    selectedIndex: PropTypes.number.isRequired,
    selectedItemScale: PropTypes.number,
    swipeSpeedMultiplier: PropTypes.number,
    theta: PropTypes.number.isRequired,
    visibilityPadding: PropTypes.number,
}
