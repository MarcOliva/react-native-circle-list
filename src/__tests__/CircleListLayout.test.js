import React from 'react'
import renderer from 'react-test-renderer'
import { CircleListLayout } from '../CircleListLayout'

// TODO: Add more test cases

test('renders correctly', () => {
    const tree = renderer
        .create(
            <CircleListLayout
                calcHeight={jest.fn()}
                displayData={[]}
                initialRotationOffset={0}
                keyExtractor={jest.fn(() => Math.random())}
                panHandlers={{}}
                radius={0}
                renderItem={jest.fn()}
                rotationOffset={0}
                selectedIndex={0}
                selectedItemScale={0}
                theta={0}
            />
        )
        .toJSON()
    expect(tree).toMatchSnapshot()
})
