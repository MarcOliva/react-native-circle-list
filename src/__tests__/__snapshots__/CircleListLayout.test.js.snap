// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    Array [
      Object {
        "alignSelf": "stretch",
        "flexDirection": "row",
        "justifyContent": "center",
        "position": "relative",
      },
      Object {
        "height": undefined,
      },
      undefined,
    ]
  }
/>
`;
